Wransom/halo fix

wransom-TT opened the pull request:
### Ticket
This PR addresses a bug fix:
https://github.com/tenstorrent/tt-metal/issues/20039

### Problem description
Conv hung for some inputs when in place halo was used.

### What's changed
This has been fixed by updating the number of cores used by the semaphores in the in place halo kernel, additionally a atomic barrier has been added to ensure all atomic operations are complete before kernel exit.
### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes:
https://github.com/tenstorrent/tt-metal/actions/runs/14316112175
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes:
https://github.com/tenstorrent/tt-metal/actions/runs/14316130876
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes:
https://github.com/tenstorrent/tt-metal/actions/runs/14316139222
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes:
https://github.com/tenstorrent/tt-metal/actions/runs/14316136699
- [x] [Nightly L2](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) CI passes:
https://github.com/tenstorrent/tt-metal/actions/runs/14316230007
https://github.com/tenstorrent/tt-metal/actions/runs/14316145738
- [x] [Frequent model](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml) CI passes:
https://github.com/tenstorrent/tt-metal/actions/runs/14316142051
- [x] New/Existing tests provide coverage for changes

wransom-TT committed:
#0: fix for in place halo num cores

wransom-TT committed:
#0: atomic barrier

s-jovic commented on ttnn/cpp/ttnn/operations/sliding_window/halo/device/kernels/dataflow/halo_gather_in_place.cpp:
```diff
@@ -303,4 +299,5 @@ void kernel_main() {
 
     noc_async_read_barrier();
     noc_async_write_barrier();
+    noc_async_atomic_barrier();
```
Can we remove dprint on line 248?

wransom-TT replied:
Unfortunately removing the DPRINT does not resolve the problem, this barrier seems to be necessary.

s-jovic reviewed (commented):
Can you remove pipelines from the description that weren't actually ran? Or uncheck

Also, can you run this pipeline as most of conv tests have been removed here: https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml

wransom-TT commented on ttnn/cpp/ttnn/operations/sliding_window/halo/device/untilize_with_halo_v2_program_factory.cpp:
```diff
@@ -461,14 +459,6 @@ operation::ProgramWithCallbacks inplace_untilize_with_halo_multi_core_v2(
     }
 
     // Compute core data and create semaphore
```
will remove!

wransom-TT commented:
@s-jovic All pipelines which are checked in the description (which includes the new L2 tests) were in fact ran, I have added links to the corresponding results to make this more clean, but will have to re-run them all anyway since I removed the DPRINT and comment!

wransom-TT committed:
#0: DPRINT comment

s-jovic reviewed (approved):
Ah sorry about that - I searched for your branch in the Actions to check if the Nightly L2 pipeline was ran, and saw only All post commit pipelines - but I was searching just within the post commit pipelines..
